### ISSUE TITLE
Performance - what is the expected throughput?

### ISSUE THREAD

**Issue by SystemParadox** [mysqljs/mysql#1796]
With a recent increase in usage of our application we have noticed that we are starting to have performance issues, with the node process using a large amount of CPU time.

A bit of profiling reveals that the app spends the large majority of its time in `RowDataPacket.parse`. This sounds reasonable enough.

However, I don't consider our app to be *that* busy, so I was wondering what the expected throughput is for this library? At the moment I would estimate our throughput to be about 16,000 rows per second. Maybe 10-20 MBits/second?

Thanks.

**Comment by sidorares** [comment]
depends heavily on the row size as well ( number of fields and type/size of the field itself )
simple optimisation might be check if you selecting only required fields

Are you using orm or build queries manually?

**Comment by SystemParadox** [comment]
Yes I realise it's dependent on row size. I'll see if I can get a more accurate estimate for this.

Our data layer is more like an ORM than anything. It's subscription-based, so the server polls the database every 5 seconds or so and sends only the changes to the clients. It's doing a fair amount of work, but we've got other similar projects using a Java platform and they hardly seem to require any processing time - CPU usage is usually less than 5%. But for this nodeJS application the CPU usage is more like 70-100% all the time.

At the moment I'm mainly just trying to get some idea of what to expect so I can see if we're even in the right ballpark. I'm trying to work out whether:

1. We've done something stupid in our app
2. This library could use some optimisation
3. It's a limitation of javascript
4. We're just processing too much data, and we wouldn't get any better performance even if we were using C

Thanks.

**Comment by sidorares** [comment]
Unless it's 1. (we've all done this :) ) 2-4 are equally possible. V8 is fast but on cpu bound tasks still 2-3 times slower than similar Java code

For tasks like yours mysql2 library might be good option (I'm main author so take this advice with a grain of salt, but worth trying. I'll be very interested to see difference in perf numbers)

**Comment by SystemParadox** [comment]
Thanks for that. I found a useful benchmark at https://github.com/Sannis/node-mysql-benchmarks (I managed to get it running for the latest versions with a few tweaks).

It seems that mysql2 performance is maybe 1.3x-1.6x compared to this library. Might be worth switching. We've made some changes to reduce out dependence on polling so we've vastly reduced the database load now.

Interestingly a pure C benchmark is only a 2-3x performance increase. It's nice to know that the difference isn't so big :)
